lib/request-handler.js

```diff
@@ -63,94 +63,100 @@ module.exports = function processRequest (options, request, response) {
     };
 
     let shouldWriteHead = true;
-    let index = 0;
 
-    contextPromise.then((context) => {
+    const contentLengthStream = new ContentLengthStream((contentLength) => {
+        this.emit('end', request, contentLength);
+    });
 
-        const contentLengthStream = new ContentLengthStream((contentLength) => {
-            this.emit('end', request, contentLength);
-        });
+    const handleError = (err) => {
+        this.emit('error', request, err);
+        if (shouldWriteHead) {
+            shouldWriteHead = false;
+            let statusCode = 500;
+            if (err.code === TEMPLATE_NOT_FOUND) {
+                statusCode = 404;
+            }
 
-        const resultStream = new StringifierStream((tag) => {
-            const { placeholder, name, } = tag;
-            if (placeholder === 'pipe') {
-                return pipeDefinition(pipeInstanceName);
+            response.writeHead(statusCode, responseHeaders);
+            // To render with custom error template
+            if (typeof err.presentable === 'string') {
+                response.end(`${err.presentable}`);
+            } else {
+                response.end();
             }
+        } else {
+            contentLengthStream.end();
+        }
+    };
+
+    Promise.all([
+        templatePromise,
+        contextPromise
+    ]).then(([ parsedTemplate, context ]) => {
+        const blockStream = buildBlock(request, Object.assign({
+            index: 0,
+            asyncStream
+        }, options, context));
+
+
+        FRAGMENT_EVENTS.forEach((eventName) => {
```

Comment:
assuming these events are for fragments. Could you explain me why it is required here?

Reply:
It's part of the tailor API (fragment:start, fragment:end, etc. events.) for fragments. It was there before (almost exact same code), difference is that now instead of subscribing to the fragment directly we subscribe to the "block" events (it propagates them), cause fragments are not directly exposed to the request handler.

Does this make sense?

Reply:
Yeah cool, looks a bit bad that we need to handle it twice.